broker/src/main/java/io/zeebe/broker/system/partitions/impl/StateControllerImpl.java

```diff
@@ -0,0 +1,312 @@
+    snapshot.take(
+        snapshotDir -> {
+          if (db == null) {
+            LOG.error("Expected to take a snapshot, but no database was opened");
+            return false;
+          }
+
+          LOG.debug("Taking temporary snapshot into {}.", snapshotDir);
+          try {
+            db.createSnapshot(snapshotDir.toFile());
+          } catch (final Exception e) {
+            LOG.error("Failed to create snapshot of runtime database", e);
+            return false;
+          }
+
+          return true;
+        });
+  }
+
+  @Override
+  public void onNewSnapshot(final PersistedSnapshot newPersistedSnapshot) {
+    // replicate snapshots when new snapshot was committed
+    try (final var snapshotChunkReader = newPersistedSnapshot.newChunkReader()) {
+      while (snapshotChunkReader.hasNext()) {
+        final var snapshotChunk = snapshotChunkReader.next();
+        replication.replicate(snapshotChunk);
+      }
+    }
+  }
+
+  /**
+   * This is called by the snapshot replication implementation on each snapshot chunk
+   *
+   * @param snapshotChunk the chunk to consume
+   */
+  private void consumeSnapshotChunk(final SnapshotChunk snapshotChunk) {
+    final String snapshotId = snapshotChunk.getSnapshotId();
+    final String chunkName = snapshotChunk.getChunkName();
+
+    final ReplicationContext context =
+        receivedSnapshots.computeIfAbsent(
+            snapshotId,
+            id -> {
+              final var startTimestamp = System.currentTimeMillis();
+              final ReceivedSnapshot transientSnapshot =
+                  store.newReceivedSnapshot(snapshotChunk.getSnapshotId());
+              return newReplication(startTimestamp, transientSnapshot);
+            });
+    if (context == INVALID_SNAPSHOT) {
+      LOG.trace(
+          "Ignore snapshot chunk {}, because snapshot {} is marked as invalid.",
+          chunkName,
+          snapshotId);
+      return;
+    }
+
+    final var receivedSnapshot = context.getReceivedSnapshot();
+    try {
+      if (receivedSnapshot.apply(snapshotChunk)) {
+        validateWhenReceivedAllChunks(snapshotChunk, context);
+      } else {
+        markSnapshotAsInvalid(context, snapshotChunk);
+      }
+    } catch (final IOException e) {
+      LOG.error("Unexepected error on writing the received snapshot chunk {}", snapshotChunk, e);
+      markSnapshotAsInvalid(context, snapshotChunk);
+    }
+  }
+
+  private void markSnapshotAsInvalid(
+      final ReplicationContext replicationContext, final SnapshotChunk chunk) {
+    replicationContext.getReceivedSnapshot().abort();
+    receivedSnapshots.put(chunk.getSnapshotId(), INVALID_SNAPSHOT);
+    metrics.decrementCount();
+  }
+
+  private void validateWhenReceivedAllChunks(
+      final SnapshotChunk snapshotChunk, final ReplicationContext context) {
+    final int totalChunkCount = snapshotChunk.getTotalCount();
+
+    if (context.incrementCount() == totalChunkCount) {
```

Comment:
There is an implicit assumption here, that all chunks return the same total count. This might not be the case if an attacker / non compliant broker modifies a snapshot chunk. You could capture the total count of the first chunk and then compare it on each subsequent chunk as an additional sanity check.

Reply:
Ok so like we do with the snapshot checksum